js/views/file_input_view.js

```diff
@@ -1,236 +1,314 @@
-                units: this.model.units
-            };
-        }
+  'use strict';
+
+  window.Whisper = window.Whisper || {};
+
+  const { MIME } = window.Signal.Types;
+
+  Whisper.FileSizeToast = Whisper.ToastView.extend({
+    templateName: 'file-size-modal',
+    render_attributes() {
+      return {
+        'file-size-warning': i18n('fileSizeWarning'),
+        limit: this.model.limit,
+        units: this.model.units,
+      };
+    },
+  });
+  Whisper.UnsupportedFileTypeToast = Whisper.ToastView.extend({
+    template: i18n('unsupportedFileType'),
+  });
+
+  function makeThumbnail(size, objectUrl) {
+    return new Promise(((resolve, reject) => {
+      const img = document.createElement('img');
+      img.onerror = reject;
+      img.onload = () => {
+        // using components/blueimp-load-image
+
+        // first, make the correct size
+        let canvas = loadImage.scale(img, {
+          canvas: true,
+          cover: true,
+          maxWidth: size,
+          maxHeight: size,
+          minWidth: size,
+          minHeight: size,
+        });
+
+        // then crop
+        canvas = loadImage.scale(canvas, {
+          canvas: true,
+          crop: true,
+          maxWidth: size,
+          maxHeight: size,
+          minWidth: size,
+          minHeight: size,
+        });
+
+        const blob = window.dataURLToBlobSync(canvas.toDataURL('image/png'));
+
+        resolve(blob);
+      };
+      img.src = objectUrl;
+    }));
+  }
+
+  function makeVideoScreenshot(objectUrl) {
+    return new Promise(((resolve, reject) => {
+      const video = document.createElement('video');
+
+      function capture() {
+        const canvas = document.createElement('canvas');
+        canvas.width = video.videoWidth;
+        canvas.height = video.videoHeight;
+        canvas.getContext('2d').drawImage(video, 0, 0, canvas.width, canvas.height);
+
+        const image = window.dataURLToBlobSync(canvas.toDataURL('image/png'));
+
+        video.removeEventListener('canplay', capture);
+
+        resolve(image);
+      }
+
+      video.addEventListener('canplay', capture);
+      video.addEventListener('error', (error) => {
+        console.log(
+          'makeVideoThumbnail error',
+          error && error.stack ? error.stack : error
+        );
+        reject(error);
+      });
+
+      video.src = objectUrl;
+    }));
+  }
+
+  function makeObjectUrl(data, contentType) {
+    const blob = new Blob([data], {
+      type: contentType,
     });
-    Whisper.UnsupportedFileTypeToast = Whisper.ToastView.extend({
-        template: i18n('unsupportedFileType')
+    return URL.createObjectURL(blob);
+  }
+
+  function blobToArrayBuffer(blob) {
+    return new Promise((resolve, reject) => {
+      const fileReader = new FileReader();
+
+      fileReader.onload = e => resolve(e.target.result);
+      fileReader.onerror = reject;
+      fileReader.onabort = reject;
+
+      fileReader.readAsArrayBuffer(blob);
     });
+  }
```

Comment:
Minor: Could we use `blobToArrayBuffer` from `blob-util`?

Reply:
I see that `modules/types/attachment.js` is pulling it in from `blob-util` today. I don't think we'll use that same function here until we convert to CommonJS.